### PR TITLE
Reminder form

### LINKED ISSUE #71: Adding Reminder Form and Tests
- [x] adding reminder form into the event creation form
- [x] adding reminder updates into the update event form
- [x] adding tests to both scenarios

### PR DESCRIPTION
**Pull request by MortarDefender**
- added reminder form
- added the form to the event creation
- added tests
- add a reminder message to tell when the event related to it from the
  reminder time
- added the option to create a reminder with the date-time as none

- fix: #71

### CODE REVIEW

File: events/views.py [hunk 1]
```diff
@@ -1,47 +1,77 @@
...
+        if event_form.is_valid() and reminder_form.is_valid():
+            event = event_form.save()
 
-        if form.is_valid():
-            form.save()
+            participant = EventParticipant.objects.get(user_id=request.user, event_id=event)
+            reminder = reminder_form.save(commit=False)
 
-            title = form.cleaned_data.get('title')
+            if reminder.date_time is not None:
```
Comment (YaelD) [inline events/views.py]:
```suggestion
            if reminder.date_time:
```

File: events/views.py [hunk 2]
```diff
@@ -1,47 +1,77 @@
...
-        form = EventUpdateForm(request.POST, user_id=request.user, instance=event_instance)
+        event_form = EventUpdateForm(request.POST, user_id=request.user, instance=event_instance)
+        reminder_form = ReminderUpdateForm(request.POST, instance=reminder_instance)
+
+        if event_form.is_valid() and reminder_form.is_valid():
+            event = event_form.save()
 
-        if form.is_valid():
-            form.save()
+            if reminder_form.instance.date_time is not None:
```
Comment (YaelD) [inline events/views.py]:
```suggestion
            if reminder_form.instance.date_time:
```

File: events/views.py [hunk 3]
```diff
@@ -1,47 +1,77 @@
...
+                reminder = reminder_form.save(commit=False)
+                reminder.messages = convert_time_delta(event.date_time_start - reminder.date_time)
+                reminder.participant_id = participant
+                reminder.save()
+            else:
+                if reminder_instance is not None:
```
Comment (YaelD) [inline events/views.py]:
```suggestion
                if reminder_instance:
```

File: main/utilities.py [hunk 4]
```diff
@@ -8,7 +8,7 @@ def send_mail_notification(title, message, receiver_email):
 
 
 def send_site_notification(message, receiver_id):
-    """ notifiy the receiver id the message given using the site notification """
+    """ notify the receiver id the message given using the site notification """
```
Comment (YaelD) [inline main/utilities.py]: This sentence is not clear enough. Maybe you meant:
```suggestion
    """ notify the receiver by his id with the message given by the site notification """
```

File: reminders/class_tests/create_reminder_tests.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    @pytest.fixture
+    def get_event_participant(self):
+        return EventParticipant.objects.get(id=1)
+
+    @pytest.fixture
+    def valid_event_data(self):
+        return {'title': "test event", 'date_time_start': DATE_TIME_START,
+                'date_time_end': DATE_TIME_END, 'color': Colors.BLACK,
+                'date_time': REMINDER_DATE_TIME, 'method': ReminderType.EMAIL}
```
Comment (YaelD) [inline reminders/class_tests/create_reminder_tests.py]: In my opinion you can use conftest.py file (which providing constants and fixtuers to entire directory), for the reminder app and put there all the constants and fixtures. In this way, you can use all the constants and fixtures in each test file in the app without importing them. 
I think that it can make our code more readable and maintainable 😄 

Here is link to conftest.py documentation:
https://docs.pytest.org/en/6.2.x/fixture.html#conftest-py-sharing-fixtures-across-multiple-files

Comment (YaelD) [inline reminders/class_tests/create_reminder_tests.py:51]: In my opinion, you can use conftest.py file (which providing constants and fixtuers to entire directory), for the reminder app and put there all the constants and fixtures. In this way, you can use all the constants and fixtures in each test file in the app without importing them. 
I think it can make our code more readable and maintainable 😄 

Here is link to conftest.py documentation:
https://docs.pytest.org/en/6.2.x/fixture.html#conftest-py-sharing-fixtures-across-multiple-files

File: reminders/class_tests/reminders_tests.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+def test_new_reminder_0(reminder_0, participant0):  # noqa: F811
+    assert reminder_0.participant_id == participant0
+    assert reminder_0.method == METHOD_0
+    assert reminder_0.messages == MESSAGES_0
+    assert reminder_0.date_time == TIME_0
+
+
+@pytest.fixture
+def persist_reminder(reminder_0):
+    return save_reminder(reminder_0)
```
Comment (YaelD) [inline reminders/class_tests/reminders_tests.py:40]: same ^^

File: reminders/class_tests/update_reminder_tests.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    @pytest.fixture
+    def get_event_participant(self):
+        return EventParticipant.objects.get(id=1)
+
+    @pytest.fixture
+    def valid_event_data(self):
+        reminder = Event.objects.get(id=1)
+        return {'title': reminder.title, 'date_time_start': reminder.date_time_start,
+                'date_time_end': reminder.date_time_end, 'color': Colors.RED,
+                'date_time': REMINDER_DATE_TIME, 'method': ReminderType.EMAIL}
```
Comment (YaelD) [inline reminders/class_tests/update_reminder_tests.py]: same^^

File: reminders/class_tests/reminders_tests.py (hunk 6 above)
Comment (YaelD) [inline reminders/class_tests/reminders_tests.py:40]: same^^

File: reminders/class_tests/reminders_tests.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+JOIN_MEETING = 'Joined Meeting in {} minutes'
+
+EXIST_REMINDER_ERROR = 'reminder already exists'
+PAST_DATE_TIME_ERROR = "date time should be bigger than the current date_time"
+
+
+@pytest.fixture
+def reminder_0(participant0):  # noqa: F811
+    return Reminder(participant_id=participant0, method=METHOD_0, messages=MESSAGES_0, date_time=TIME_0)
+
```
Comment (YaelD) [inline reminders/class_tests/reminders_tests.py:29]: same^^

File: reminders/class_tests/reminders_tests.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    reminder.participant_id.user_id.save()
+    reminder.participant_id.event_id.save()
+    reminder.participant_id.save()
+    reminder.save()
+    return reminder
+
+
+@pytest.fixture(autouse=True)
+def email_backend_setup():
+    settings.EMAIL_BACKEND = 'django.core.mail.backends.locmem.EmailBackend'
```
Comment (YaelD) [inline reminders/class_tests/reminders_tests.py:57]: same^^

File: reminders/class_tests/reminders_tests.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
```
Comment (YaelD) [inline reminders/class_tests/reminders_tests.py]: same^^

Comment (YaelD) [inline reminders/class_tests/reminders_tests.py:58]: same^^

File: reminders/class_tests/update_reminder_tests.py (hunk 7 above)
Comment (YaelD) [inline reminders/class_tests/update_reminder_tests.py:54]: same^^

Comment (YaelD) [inline reminders/class_tests/update_reminder_tests.py:54]: same^^

File: reminders/models.py [hunk 11]
```diff
@@ -17,8 +17,9 @@ def get_next_reminder(self):
 
 
 def validate_date(date_time):
-    if timezone.now() > date_time:
-        raise ValidationError('date time should be bigger than the current date_time')
+    if date_time is not None:
```
Comment (YaelD) [inline reminders/models.py]:
```suggestion
    if date_time:
```

Comment (YaelD) [inline reminders/models.py:20]:
```suggestion
    if date_time:
```

File: reminders/scheduler.py [hunk 12]
```diff
@@ -40,9 +40,9 @@ class UserAlertScheduler():
 
     def __new__(self):
         """
-        __new__ is used becuse we want to make the scheduler a singletone
+        __new__ is used becurse we want to make the scheduler a singletone
```
Comment (YaelD) [inline reminders/scheduler.py]:
```suggestion
        __new__ is used because we want to make the scheduler a singletone
```

Comment (YaelD) [inline reminders/scheduler.py:43]:
```suggestion
        __new__ is used because we want to make the scheduler a singletone
```

File: reminders/scheduler.py [hunk 13]
```diff
@@ -187,23 +187,25 @@ def __remove_alert(reminder_id):
...
-                    UserAlertScheduler.__current_timer = None
-
-                    UserAlertScheduler.__logger.debug(
-                        f"removed the current reminder. last reminer: {Reminder.objects.get(id=reminder_id)}"
-                    )
-                else:
-                    UserAlertScheduler.__logger.warning(
-                        f"the remidner {UserAlertScheduler.__current_reminder} is the current remidner but has no timer"
-                    )
+                if UserAlertScheduler.__current_timer is not None:
```
Comment (YaelD) [inline reminders/scheduler.py]:
```suggestion
                if UserAlertScheduler.__current_timer:
```

Comment (YaelD) [inline reminders/scheduler.py:190]:
```suggestion
                if UserAlertScheduler.__current_timer:
```

File: users/views.py [hunk 14]
```diff
@@ -100,11 +100,12 @@ def main_page(request, date=None):
     current_date = (datetime.now().year, datetime.now().month, datetime.now().day)
     current_events = Event.objects.filter(eventparticipant__user_id=request.user.id)
     table_measurements = get_table_measurements(current_events)
+    year = year if year is not None else current_date[0]
```
Comment (YaelD) [inline users/views.py]:
```suggestion
    year = year if year else current_date[0]
```

Comment (YaelD) [inline users/views.py:103]:
```suggestion
    year = year if year else current_date[0]
```

**Review by YaelD — CHANGES_REQUESTED** [review]

**Review by YaelD — CHANGES_REQUESTED** [review]

File: reminders/class_tests/create_reminder_tests.py (hunk 5 above)
Comment (MortarDefender) [reply reminders/class_tests/create_reminder_tests.py]: Great Catch, 💪 
I will add a new file conftest.py to the app directory so 
that all the tests will have a global area to share fixtures and data. 

done. 😄

File: reminders/class_tests/update_reminder_tests.py (hunk 7 above)
Comment (MortarDefender) [reply reminders/class_tests/update_reminder_tests.py]: done. thanks 😄

File: reminders/class_tests/reminders_tests.py (hunk 10 above)
Comment (MortarDefender) [reply reminders/class_tests/reminders_tests.py]: done. thanks 😄

File: events/views.py (hunk 1 above)
Comment (MortarDefender) [reply events/views.py]: done. thanks 😄

File: users/views.py (hunk 14 above)
Comment (MortarDefender) [reply users/views.py]: done. thanks 😄

File: reminders/scheduler.py (hunk 13 above)
Comment (MortarDefender) [reply reminders/scheduler.py]: done. thanks 😄

File: reminders/scheduler.py (hunk 12 above)
Comment (MortarDefender) [reply reminders/scheduler.py]: done. thanks 😄

File: reminders/models.py (hunk 11 above)
Comment (MortarDefender) [reply reminders/models.py]: done. thanks 😄

File: main/utilities.py (hunk 4 above)
Comment (MortarDefender) [reply main/utilities.py]: done. thanks 😄

File: events/views.py (hunk 3 above)
Comment (MortarDefender) [reply events/views.py]: done. thanks 😄

File: events/views.py (hunk 2 above)
Comment (MortarDefender) [reply events/views.py]: done. thanks 😄

**Review by YaelD — DISMISSED** [review]
Looks great! 😄 
and nice use of conftest 👍

File: reminders/class_tests/create_reminder_tests.py (hunk 5 above)
Comment (YaelD) [reply reminders/class_tests/create_reminder_tests.py]: Great! 😄

File: main/utilities.py [hunk 15]
```diff
@@ -22,3 +22,45 @@ def send_reminder_email(message, receiver_email):
...
+        if result[field_name] == 0:
+            continue
+
+        converted_result += f"{field_name}: {result[field_name]} "
+
+    return converted_result
+
+
+def check_and_get_time(time, result, field, time_delta):
+    """ get a time variable that holds amount of the time wanted to check
```
Comment (RonHachmon) [inline main/utilities.py]: can you please make this comment more readable in general

File: reminders/class_tests/create_reminder_tests.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        {'participant_id': None, 'date_time': pytest.invalid_date_time,
+         'messages': pytest.message, 'method': pytest.valid_method
+         }
+    ], ids=[
+        "method is none",
+        "method is not from ReminderType",
+        "date is before the current time"
+        ]
+    )
+    def test_create_reminder(self, get_event_participant, invalid_user_credentials):
```
Comment (RonHachmon) [inline reminders/class_tests/create_reminder_tests.py]: please change the method name to creating invalid reminder, also if possible make also create a test for a valid form

File: reminders/class_tests/update_reminder_tests.py [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+        {'participant_id': None, 'date_time': pytest.invalid_date_time,
+         'messages': pytest.message, 'method': pytest.valid_method
+         }
+    ], ids=[
+        "method is none",
+        "method is not from ReminderType",
+        "date is before the current time"
+        ]
+    )
+    def test_create_reminder(self, get_event_participant, invalid_user_credentials):
```
Comment (RonHachmon) [inline reminders/class_tests/update_reminder_tests.py:28]: please change name to invalid reminder

File: users/templates/base.html [hunk 18]
```diff
@@ -4,7 +4,7 @@
     <meta name="viewport" content="width=device-width, initial-scale=1">
     <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
     <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
-    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
+    <!-- <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css"> -->
```
Comment (RonHachmon) [inline users/templates/base.html]: is this comment necessary?

File: events/templates/events/create_event.html [hunk 19]
```diff
@@ -21,12 +20,15 @@
                     <form method="POST">
                         {% csrf_token %}
 
-                        {{ form }}
+                        {{ event_form }}
+                        <br><br>
+
+                        <h3 style="color:white;"> Reminder: </h3>
+                        {{ reminder_form }}
```
Comment (RonHachmon) [inline events/templates/events/create_event.html]: IMHO i think we should make the reminder form hidden, and only after a user click 'add reminder' button to make it visible.
similiar to what i did on the submit button on the #80,
because by how it looks now, it feels like you have to make a reminder.

**Review by RonHachmon — CHANGES_REQUESTED** [review]

File: users/templates/base.html (hunk 18 above)
Comment (MortarDefender) [reply users/templates/base.html]: yes

File: events/templates/events/create_event.html (hunk 19 above)
Comment (MortarDefender) [reply events/templates/events/create_event.html]: done. 
created a bell icon that when clicked the form will show itself,
and when clicking again on the bell icon it will hide the form

File: users/templates/base.html (hunk 18 above)
Comment (MortarDefender) [reply users/templates/base.html]: this line can interrupt in some cases with other stylesheets,
for now I uncommented it

File: reminders/class_tests/update_reminder_tests.py (hunk 17 above)
Comment (MortarDefender) [reply reminders/class_tests/update_reminder_tests.py:28]: done as you can see in the above comment

File: reminders/class_tests/create_reminder_tests.py (hunk 16 above)
Comment (MortarDefender) [reply reminders/class_tests/create_reminder_tests.py]: done, added a new method, and change the old one name.

File: main/utilities.py (hunk 15 above)
Comment (MortarDefender) [reply main/utilities.py]: done, I guess...
see if there is an improvement

**Review by RonHachmon — DISMISSED** [review]
Great work,
approved 👍

**Review by YaelD — DISMISSED** [review]

File: events/static/events/create_event.css [hunk 20]
```diff
@@ -163,13 +164,15 @@ input[type="checkbox"] + span {
 @media (max-width: 524px) {
   .container-shadow {
     width: 400px;
+    /* width: 80%; */
```
Comment (kobihk) [inline events/static/events/create_event.css]: maybe you can remove it?

File: events/static/events/create_event.css [hunk 21]
```diff
@@ -163,13 +164,15 @@ input[type="checkbox"] + span {
...
     margin-top: 60px;
     left: 50%;
     margin-left: -200px;
   }
 
   .container {
     width: 400px;
+    /* width: 80%; */
```
Comment (kobihk) [inline events/static/events/create_event.css]: same ^^

**Review by kobihk — CHANGES_REQUESTED** [review]

File: events/static/events/create_event.css (hunk 20 above)
Comment (MortarDefender) [reply events/static/events/create_event.css]: done. thanks 😄

File: events/static/events/create_event.css (hunk 21 above)
Comment (MortarDefender) [reply events/static/events/create_event.css]: done. thanks 😄

**Review by RonHachmon — APPROVED** [review]

**Review by YaelD — APPROVED** [review]

**Review by kobihk — APPROVED** [review]
